### ISSUE TITLE
Pilon error: "Inappropriate call if not paired read" when using single end data 

### ISSUE THREAD

**Issue by swlong** [rrwick/Unicycler#56]
I was thinking at first this was a possible issue with Pilon but seeing the error it looks like an issue on how pilon is being called when short read data is being used. 

When using Unicycler with ONT long reads and a single end Illumina library, first two rounds of Pilon complete fine with "--fix bases" option. First round makes one change, second round finds no changes, third round runs with "--fix all" and that seems to lead to the crash. I can't find any documentation describing if pilon has requirements for single or paired end data for any of the --fix options, but that's the only difference I see. 

The Pilon crash doesn't prevent Unicycler from finishing. I have the whole log available if needed, pasting the relevant portion here.

```
Pilon polish round 3
  bowtie2-build 3_polish_input.fasta 3_polish_input.fasta
  bowtie2 --local --very-sensitive-local --threads 12 -I 0 -X 1000 -x 3_polish_input.fasta -S 3_alignments.sam -U /home/manager/HG/Foo_1.fastq
  samtools sort -@ 12 -o 3_alignments.bam -O bam -T temp 3_alignments.sam
  samtools index 3_alignments.bam
  java -jar /home/manager/pilon/pilon-1.22.jar --genome 3_polish_input.fasta --frags 3_alignments.bam --changes --output 3_pilon --outdir /home/manager/Uni_HGHybridv2/pilon_polish --fix all
Unable to polish assembly using Pilon: Pilon encountered an error:
Pilon version 1.22 Wed Mar 15 16:38:30 2017 -0400
Genome: 3_polish_input.fasta
Fixing snps, indels, gaps, local
Input genome size: 1853100
Scanning BAMs
3_alignments.bam: 18046075 reads, 0 filtered, 17915535 mapped, 0 proper, 0 stray, Unpaired 100% 50+/-0, max 51
Processing 1:1-1853100
frags 3_alignments.bam: coverage 290
Total Reads: 17915535, Coverage: 290, minDepth: 29
Confirmed 1849048 of 1853100 bases (99.78%)
Corrected 0 snps; 0 ambiguous bases; corrected 0 small insertions totaling 0 bases, 0 small deletions totaling 0 bases
Large collapsed region: 1:1528003-1545888 size 17886
Large collapsed region: 1:1637611-1650580 size 12970
# Attempting to fix local continuity breaks
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.simontuffs.onejar.Boot.run(Boot.java:340)
	at com.simontuffs.onejar.Boot.main(Boot.java:166)
Caused by: java.lang.IllegalStateException: Inappropriate call if not paired read
	at htsjdk.samtools.SAMRecord.requireReadPaired(SAMRecord.java:648)
	at htsjdk.samtools.SAMRecord.getFirstOfPairFlag(SAMRecord.java:706)
	at org.broadinstitute.pilon.BamFile$MateMap.addRead(BamFile.scala:224)
	at org.broadinstitute.pilon.BamFile$MateMap$$anonfun$addReads$1.apply(BamFile.scala:220)
	at org.broadinstitute.pilon.BamFile$MateMap$$anonfun$addReads$1.apply(BamFile.scala:220)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.broadinstitute.pilon.BamFile$MateMap.addReads(BamFile.scala:220)
	at org.broadinstitute.pilon.BamFile$MateMap.<init>(BamFile.scala:218)
	at org.broadinstitute.pilon.BamFile.recruitFlankReads(BamFile.scala:339)
	at org.broadinstitute.pilon.GapFiller$$anonfun$recruitReadsOfType$1.apply(GapFiller.scala:367)
	at org.broadinstitute.pilon.GapFiller$$anonfun$recruitReadsOfType$1.apply(GapFiller.scala:366)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.broadinstitute.pilon.GapFiller.recruitReadsOfType(GapFiller.scala:366)
	at org.broadinstitute.pilon.GapFiller.recruitFrags(GapFiller.scala:375)
	at org.broadinstitute.pilon.GapFiller.recruitLocalReads(GapFiller.scala:389)
	at org.broadinstitute.pilon.GapFiller.recruitReads(GapFiller.scala:391)
	at org.broadinstitute.pilon.GapFiller.assembleAcrossBreak(GapFiller.scala:51)
	at org.broadinstitute.pilon.GapFiller.fixBreak(GapFiller.scala:45)
	at org.broadinstitute.pilon.GenomeRegion$$anonfun$identifyAndFixIssues$4.apply(GenomeRegion.scala:383)
	at org.broadinstitute.pilon.GenomeRegion$$anonfun$identifyAndFixIssues$4.apply(GenomeRegion.scala:381)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.broadinstitute.pilon.GenomeRegion.identifyAndFixIssues(GenomeRegion.scala:381)
	at org.broadinstitute.pilon.GenomeFile$$anonfun$processRegions$4.apply(GenomeFile.scala:119)
	at org.broadinstitute.pilon.GenomeFile$$anonfun$processRegions$4.apply(GenomeFile.scala:108)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:972)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:49)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:51)
	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:969)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:152)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:443)
	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)


```
Best,
Wesley

**Comment by shyamrallapalli** [comment]
I had this issue today
It seems this is a [reported issue with pilon](https://github.com/broadinstitute/pilon/issues/51)
Might be fixed in the future versions.
For the moment best thing would be to pass unpaired reads 
as a seperate bam file to pilon with ```--unpaired <unpaired.bam>```

I have tested this to work by providing separare paired and unpaired bams
```
java -jar pilon1.22.jar --genome 4_polish_input.fasta --frags pair_4_alignments.bam --unpaired unpair_4_alignments.bam --changes --output 4_pilon --outdir pilon_polish --fix all
```

**Comment by swlong** [comment]
In looking at the pilon usage notes, it seems to indicate that --frags is used for passing paired-end read alignments while --unpaired is used for passing unpaired or single-end alignments. There is also some indication that starting with pilon 1.17 --bam will determine which type of read alignment is being passed (paired, unpaired, or long mate pair reads).

So perhaps the next version of unicycler could either sub in --unpaired for --frags when using unpaired read alignments, or perhaps using the --bam option instead of --frags will solve the issue, if Pilon 1.17 is a required minimum version.  

I've rerun the pilon steps using --unpaired instead of --frags and it seems to more or less be functional, although the "fix all" step is identifying numerous "fix breaks" with NoSolution as the result. Not sure what to make of that.

**Comment by shyamrallapalli** [comment]
Hi @swlong I don't think ```--bam``` option (from pilon >=1.17) would be the fix
As I understand it is just going to check if bam is paired, mate-paired or unpaired and set the flag accordingly.

I have edited `polish_with_pilon` function in [pilon_func.py](https://github.com/rrwick/Unicycler/blob/250abc3f798eda046c0522b4de9be222f415e572/unicycler/pilon_func.py#L157)
to generate separate bam files from paired and/or unpaired inputs files (when they are available)
and use them accordingly with Pilon. 
Currently testing it, will either post the code or make a pull request.

**Comment by shyamrallapalli** [comment]
As mentioned earlier, I have made a [pull request for my changes](https://github.com/rrwick/Unicycler/pull/58)
This resolves the issue and successfully completes Pilon polishing of gaps and misassemblies.
@rrwick I have just moved bowtie and samtools commands part out and called them separately for paired and unpaired data.

**Comment by swlong** [comment]
I have only tested it with either paired or unpaired FASTQ files, but the --bam switch appears to work for my purposes. I'll leave it up to @rrwick as to which is the better approach. I'm agnostic.

**Comment by rrwick** [comment]
Thanks to both of you for helping with this one! I actually merged in S. Wesley's PR, then I noticed Shyam's PR, _then_ I noticed this issue... slightly out of order :smile:

I share Shyam's concern that Pilon may treat `--bam` files as _either_ paired or unpaired, but wouldn't properly handle both. So in the end I've gone with his approach. The changes are merged into the development branch. You're welcome to try that branch, but keep in mind that's where I try things so it does break sometimes. The fix will make it to the master branch for the next release.

Thanks again!
